### PR TITLE
[exporter/syslog] Add syslog exporter

### PR DESCRIPTION
**Pull request by kasia-kujawa**
**Description:** Adding syslog exporter implementation

**Link to tracking Issue:** #17982

**Testing:** unit tests, manual tests with syslog server

**Documentation:** please see README.md file

### CODE REVIEW

**Comment by kkujawa-sumo** [comment]
This pull request is ready for review :) 
cc: @astencel-sumo @djaglowski

File: exporter/syslogexporter/README.md
Comment (astencel-sumo) [inline exporter/syslogexporter/README.md:3]: We need a Markdown table here that lists the stability, pipeline types and distros like every component has, see example [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/3191d9cc96585002e2ab2d6a739c42e3f9e5fe2f/exporter/zipkinexporter/README.md#L3-L7).

We should probably start with [Development](https://github.com/open-telemetry/opentelemetry-collector/blob/main/README.md#development) stability, not Alpha.

File: exporter/syslogexporter/README.md
Comment (astencel-sumo) [inline exporter/syslogexporter/README.md:5]: Do we need this header? Maybe just drop it? I think the other components don't have the "About" section specifically called out.

File: exporter/syslogexporter/README.md
Comment (astencel-sumo) [inline exporter/syslogexporter/README.md:16]: > The following are a few configuration options

Does this mean these are not all of them? :sweat_smile:

```suggestion
**The following configuration options are available**:
```

File: exporter/syslogexporter/config.go
Comment (astencel-sumo) [inline exporter/syslogexporter/config.go:1]:
```suggestion
// Copyright The OpenTelemetry Authors
```

File: exporter/syslogexporter/config.go
Comment (astencel-sumo) [inline exporter/syslogexporter/config.go:60]: Do we need to validate the format of the endpoint? I don't see other components doing this, and removing this would save us from adding another dependency (github.com/THREATINT/go-net with 10 stars on GitHub).

File: exporter/syslogexporter/README.md
Comment (astencel-sumo) [inline exporter/syslogexporter/README.md:14]: The `Configuration` section only mentions the `retry_on_failure` or the `sending_queue` settings in the example. I believe these settings and their enabled status should be described in the list of all configuration settings.

File: exporter/syslogexporter/config.go
Comment (astencel-sumo) [inline exporter/syslogexporter/config.go:56]: Shouldn't this be 65535?

```suggestion
	if cfg.Port < 1 || cfg.Port > 65535 {
```

**Comment by kkujawa-sumo** [comment]
@atoulme @djaglowski I went through steps described in https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/CONTRIBUTING.md#adding-new-components and I have some doubts related to two steps:

> Run make crosslink to update intra-repository dependencies. It will add a replace directive to go.mod file of every intra-repository dependant. This is necessary for your component to be included in the contrib executable. 

I didn't see any changes after executing `make crosslink`

> All components must be included in [internal/components/](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/internal/components) and in the respective testing harnesses. To align with the test goal of the project, components must be testable within the framework defined within the folder. If a component can not be properly tested within the existing framework, it must increase the non testable components number with a comment within the PR explaining as to why it can not be tested.

Should I add syslog exporter to `internal/components/` in this pull request or it is better/expected to do it in another pull request?

File: exporter/syslogexporter/factory.go
Comment (astencel-sumo) [inline exporter/syslogexporter/factory.go:45]: This is not related to this pull request specifically, but I stumbled upon the fact that the TLS settings apparently do not have a `NewDefaultSettings` method similar to `NewDefaultRetrySettings` or `NewDefaultQueueSettings`. Perhaps it's worth adding such a method on the [TLSSetting](https://github.com/open-telemetry/opentelemetry-collector/blob/v0.73.0/config/configtls/configtls.go) struct?

This is not a problem now, because the default TLS config is equal to an empty, non-initialized struct. But if one day this changes, it would be useful to have the `NewDefaultSettings` method as a central place to change this.

I wonder what the @open-telemetry/collector-approvers think.

File: exporter/syslogexporter/syslog.go
Comment (astencel-sumo) [inline exporter/syslogexporter/syslog.go:47]: Why is this named `Syslog`? It seems a name like `sender` or `syslogSender` would be a better name, as it is concerned with sending the syslog data to a remote endpoint?

Also I don't think this struct needs to be exported (i.e. start with a capital letter)?

File: exporter/syslogexporter/config.go
Comment (kkujawa-sumo) [reply exporter/syslogexporter/config.go]: changed

File: exporter/syslogexporter/README.md
Comment (kkujawa-sumo) [reply exporter/syslogexporter/README.md]: changed

File: exporter/syslogexporter/README.md
Comment (kkujawa-sumo) [reply exporter/syslogexporter/README.md]: changed

File: exporter/syslogexporter/README.md
Comment (kkujawa-sumo) [reply exporter/syslogexporter/README.md]: changed

File: exporter/syslogexporter/config.go
Comment (kkujawa-sumo) [reply exporter/syslogexporter/config.go]: chnaged

File: exporter/syslogexporter/config.go
Comment (kkujawa-sumo) [reply exporter/syslogexporter/config.go]: changed

File: exporter/syslogexporter/README.md
Comment (kkujawa-sumo) [reply exporter/syslogexporter/README.md:16]: I added information about `retry_on_failure` and `sending_queue` to README.md

File: exporter/syslogexporter/README.md
Comment (kkujawa-sumo) [reply exporter/syslogexporter/README.md]: I changed it

Comment (astencel-sumo) [reply exporter/syslogexporter/README.md]: You probably also want to remove the `DefaultEndpoint` const from `config.go` (it's still being used in the `createDefaultConfig` function in `factory.go`).

File: exporter/syslogexporter/config.go
Comment (astencel-sumo) [inline exporter/syslogexporter/config.go:87]: This should probably be removed?

File: .github/dependabot.yml
Comment (astencel-sumo) [inline .github/dependabot.yml:1100]: Why is this being removed?

File: exporter/syslogexporter/syslog.go
Comment (astencel-sumo) [inline exporter/syslogexporter/syslog.go:1]: Please rename the file too

File: exporter/syslogexporter/exporter.go
Comment (astencel-sumo) [inline exporter/syslogexporter/exporter.go:74]: Shouldn't we expose the timeout as a configuration option?

File: .github/dependabot.yml
Comment (kkujawa-sumo) [reply .github/dependabot.yml:1100]: There was problem reporter in checks:
```
diff --git a/.github/dependabot.yml b/.github/dependabot.yml
index bce0138..775dd60 100644
--- a/.github/dependabot.yml
+++ b/.github/dependabot.yml
@@ -267,6 +267,11 @@ updates:
     schedule:
       interval: "weekly"
       day: "wednesday"
+  - package-ecosystem: "gomod"
+    directory: "/exporter/syslogexporter"
+    schedule:
+      interval: "weekly"
+      day: "wednesday"
   - package-ecosystem: "gomod"
     directory: "/exporter/tanzuobservabilityexporter"
     schedule:
@@ -1097,8 +1102,3 @@ updates:
     schedule:
       interval: "weekly"
       day: "wednesday"
-  - package-ecosystem: "gomod"
-    directory: "/testbed"
-    schedule:
-      interval: "weekly"
-      day: "wednesday"
dependabot.yml is out of date, please run "make gendependabot" and commit the changes in this PR.
Error: Process completed with exit code 1.
```
https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/4437454266/jobs/7811020637
so I run "make gendependabot" and commited the changes in this PR.

File: exporter/syslogexporter/config.go
Comment (kkujawa-sumo) [reply exporter/syslogexporter/config.go]: removed

File: exporter/syslogexporter/sender_test.go
Comment (astencel-sumo) [inline exporter/syslogexporter/sender_test.go:60]: Why do you use regular expression here instead of raw strings? Is it because the order of the fields is undefined?

Comment (kkujawa-sumo) [reply exporter/syslogexporter/sender_test.go:60]: yes, we cannot predict the order of elements in structured data

**Review by andrzej-stencel — APPROVED** [review]
This is a good start. I would like to see some improvements, but it's probably best to include them in follow-up PRs instead of blocking this one forever.

Things that come to my mind:
- Include some examples in docs - input vs. output.
- Document the default values being used when data is missing (like default priority or default facility).
- Perhaps allow users to change the defaults with the configuration? Or maybe it's not needed, users can use processors for this.
- It seems that performing a separate network request for each Resource Log is not needed - perhaps all ResourceLogs could be sent in one request?
- Also not sure if/why we need to close the network connection after each request (with the call to `sender.close()` in the `sender.sendSyslogs()` method).

I think these are all details that can be worked on in separate PRs, if needed.

Thank you @kkujawa-sumo for the great work and for your patience with me :pray: :sweat_smile:

**Comment by astencel-sumo** [comment]
@djaglowski can you take a look? Can you enable to CI workflows to see test results?

**Review by djaglowski — COMMENTED** [review]
This looks good to me. Just one outstanding question.

I think we should determine which additional work should be done before moving the component to alpha, and we should create tracking issues for those things.

One thing I would like to see is an integration test between this component and the syslog receiver, but this can be in another PR. The easiest way to do it is probably by adding it to the testbed.

File: exporter/syslogexporter/config.go
Comment (astencel-sumo) [reply exporter/syslogexporter/config.go:43]: Yeah we discussed this with Kasia, the incosistency is far from ideal. Kasia was willing to change the names in the exporter to match the receiver (so use `protocol` for the RFC and something else for the network protocol). My opinion was that the exporter's naming is actually better, which is probably highly subjective. :sweat_smile: After all, a "protocol" is just an "agreed-on form of communication" in the widest sense, so I guess the word fits both uses.

Given your feedback, I'm OK with using `protocol` for the RFC and `network` for TCP/UDP. @kkujawa-sumo what do you think?

Comment (astencel-sumo) [reply exporter/syslogexporter/config.go:43]: I stumbled upon the [Carbon receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.75.0/receiver/carbonreceiver/README.md) using `transport` name for this configuration option (TCP or UDP). Whether this is a better name, I don't know :sweat_smile:

Comment (djaglowski) [reply exporter/syslogexporter/config.go:43]: 👍 Either `network` or `transport` seems fine to me.

File: exporter/syslogexporter/factory.go
Comment (astencel-sumo) [inline exporter/syslogexporter/factory.go:29]: Shouldn't this be Development? README says it's Development, not Alpha.

File: exporter/syslogexporter/config.go
Comment (kkujawa-sumo) [reply exporter/syslogexporter/config.go]: Parameters are changed.

File: exporter/syslogexporter/factory.go
Comment (kkujawa-sumo) [reply exporter/syslogexporter/factory.go]: I updated this pull request to use the new mechanism.

**Comment by kkujawa-sumo** [comment]
@astencel-sumo @djaglowski please take a look once again ;) 
I created following issues which must be done before giving this exporter alpha status:
- https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/21245
- https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/21242
- https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/21244

**Review by djaglowski — APPROVED** [review]

File: exporter/syslogexporter/README.md
Comment (astencel-sumo) [inline exporter/syslogexporter/README.md:62]: This line is probably not needed anymore, as the link is being autogenerated together with the status table.
```suggestion
```

**Review by andrzej-stencel — APPROVED** [review]

**Comment by astencel-sumo** [comment]
Not sure where the check errors come from:
- [check-collector-module-version](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/4891495598/jobs/8785722737?pr=19647):
  Incorrect version "v0.76.1" of "go.opentelemetry.io/collector " is included in "./exporter/syslogexporter/go.mod". It should be version "v0.76.2-0.20230502195822-4df44379e094".
- [checks](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/4891495598/jobs/8785722983?pr=19647):
  'status' has invalid keys: pipelines
  'status.stability' expected a map, got 'string'

Perhaps you need to merge the latest changes from `main`? 🤔

**Comment by dmitryax** [comment]
@kkujawa-sumo please rebase, run make gotidy and update metadata.yaml to the latest schema

**Comment by kkujawa-sumo** [comment]
@dmitryax All checks passed :)  could you merge this pull request?

File: exporter/syslogexporter/README.md
Comment (djaglowski) [reply exporter/syslogexporter/README.md:61]: @kkujawa-sumo, please take note of this. It can be done in a follow up PR.

**Pull request merged**
